---
title: 'Group 1: Expression profiling by array analysis'
author: 
- Luca Visser
date: 2022
output:
  pdf_document:
    dev: png
    toc: true
    keep_tex: true
  html_document:
    df_print: paged
    toc: true
header-includes: \setlength{\emergencystretch}{3em}
---

```{r global_options, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(out.width = '60%')
knitr::opts_chunk$set(fig.width=12, fig.height=8)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r}
suppressPackageStartupMessages({
  library(affy)
  library(arrayQualityMetrics)
  library(ArrayExpress)
  library(limma)
  library(oligo)
  library(siggenes)
  library(htmltools)
  library(biomaRt)
  library(dplyr)
  library(GEOquery)
  library(annotation)
  library(annotate)
  library(affycoretools)
  library(knitr)
})
```

# General info

Platform type: Affymetrix Human Gene 1.0 ST Array [transcript (gene) version]. 

The GEO dataset with accession ID GSE36980 (expression profiling by array of human postmortal brain tissues inclusing frontal cortex, lateral cortex and hippocampus) will be analysed in a case versus control framework for differential gene expression.

# Pipeline

## Import intensity values and metadata

The corresponding GEO SOFT file is parsed into an R data structure.

```{r}
# Retrieve GEO accession
id <- "GSE36980"
gse <- getGEO(id, GSEMatrix = TRUE)
gse <- gse[[1]]
```
The supplementary files contain the raw intensity values in the form if CEL files. These must first be downloaded and subsequently unzipped, which this chunk will perform if not done manually from the browser page.

```{r eval = FALSE}
# Download and unzip CEL files
unlink(id, recursive = T, force = T)
filePath <- rownames(getGEOSuppFiles(id))
untar(filePath, exdir = paste0(id, "/data"))
for (file in list.files(paste0(id, "/data"), full.names = T)){
  gunzip(file)
}
```
Within the parsed data the experiment annotations can be found. Seeing as the RNA-seq was performed exclusively on hippocampus tissue, we wish to only read in CEL files associated with this tissue type.

```{r}
# Store annotation and filter to only keep hippocampus samples
annot <- pData(gse)
head(annot)

# Filter to only keep annotation of hippocampus samples
temp <- annot[annot$characteristics_ch1 == "tissue: Hippocampus",]
annot.filt <- temp[order(temp$title),] %>%
  mutate(disease = c(rep("AD", 8), rep("CT", 10))) %>%
  AnnotatedDataFrame()
```

```{r message = FALSE, warning = FALSE}
# Only read in CEL files associated with hippocampus
acc <- annot.filt$geo_accession
celfiles <- list.celfiles(paste0(id, "/data"), full.names = T)
cels.hippo <- vector()
for (i in acc){
  cels.hippo <- append(cels.hippo, grep(i, celfiles, value = T))
}

data.raw <- read.celfiles(cels.hippo, verbose=T, phenoData=annot.filt)
head(exprs(data.raw))
```

## Quality Control on raw data
We perform quality control on both the remaining pure raw data and log-transformed raw data. A summary of each report can be assessed by opening the "index.html"-file.

```{r eval = FALSE}
arrayQualityMetrics(data.raw,outdir="raw",force=T)
arrayQualityMetrics(data.raw,outdir="rawlog",force=T,do.logtransform=T)
```

## Preprocessing of raw data

Currently, there are 10 control samples versus 8 cases, which is an acceptable ratio. We still choose to remove sample 'GSM4764672' as this was shown in the QC to be a severe outlier, most likely due to this sample being taken at a much later date than the rest.

Background correction, normalization and summarization are subsequently performed with the 'oligo' package.

```{r}
# Remove outliers based on QC
r <- c("GSM4764672")
cels.hippo.filt <- cels.hippo[!grepl(paste(r,collapse="|"), cels.hippo)]

annot.filt <- pData(data.raw)

annot.filt <- annot.filt %>%
  filter(!(geo_accession %in% r)) %>%
  AnnotatedDataFrame()

data.raw <- read.celfiles(cels.hippo.filt, verbose=T, phenoData=annot.filt)

# RMA with oligo
data.norm <- oligo::rma(data.raw,target = "core")
```
```{r}
annot.filt <- pData(data.norm) %>%
  rename(sex = "Sex:ch1") %>%
  mutate(sex = factor(sex)) %>%
  mutate(disease = factor(disease)) %>%
  mutate(patient_id = factor(paste0(rep("p", 17), 1:17)))
plotMDS(data.norm, top = 500, col=as.double(annot.filt$disease))
```

The probeIDs must be converted into gene IDs with a library before they can be used for differential gene expression analysis.

```{r}
library(hugene10sttranscriptcluster.db)
ID <- featureNames(data.norm)

geneAcc <- hugene10sttranscriptclusterENSEMBL[ID]
rownames(data.norm) <- paste(as.character(c(1:nrow(data.norm))), geneAcc, sep=".")

hist(table(as.factor(as.character(geneAcc))), xlab = "Amount of duplicate genes", main="Histogram of uniqueness of gene symbols in dataset.")
```
```{r}
d <- exprs(data.norm)
boxplot(d)
```
```{r}
d2 <- normalizeQuantiles(d)
boxplot(d2)
humanRMA <- d
```
## Quality control on preprocessed data

```{r eval = FALSE}
QC <- arrayQualityMetrics(data.norm, outdir="rma", force=TRUE, intgroup = c("age:ch1")) #RMA produces log-transformed data
```

## Differential expression analysis with RMA preprocessed data

```{r}
head(pData(data.norm))
head(humanRMA)
```

```{r}
## Differential expression by LIMMA
design <- model.matrix(~sex+disease, data = annot.filt)
fit <- lmFit(data.norm, design)

cont.matrix <- makeContrasts(ADvsCT="diseaseCT",levels=design)
rownames(cont.matrix)[1] <- "(Intercept)"

fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)

results <- decideTests(fit2)
summary(results)
```
```{r}
# Volcano plot
volcanoplot(fit2)
# MA plot
p <- fit2$p.value < 0.05
limma::plotMA(fit2, status = p, hl.cex = 0.1)
abline(h=c(1,-1), col=c("blue", "blue"), lty=c(2,2))
```

```{r}
# DE results
LIMMAout_RMA <- topTable(fit2, adjust="BH", number=nrow(data.norm))
head(LIMMAout_RMA, n=10)
```

## Gene set analysis

```{r}
LIMMAout_filtered <- LIMMAout_RMA[LIMMAout_RMA$adj.P.Val < 0.1, ]
EntrezIDs <- mapIds(hugene10sttranscriptcluster.db, gsub('.*\\.', '', rownames(LIMMAout_filtered)), 'ENTREZID', 'ENSEMBL')

# Subset for non duplicated and mapped genes
IDs_unique <- EntrezIDs[!(duplicated(EntrezIDs) | is.na(EntrezIDs))]

goanaOut <- goana(de=IDs_unique, species="Hs", trend=T)
goanaOut <- goanaOut[order(goanaOut$P.DE, decreasing=FALSE), ]
goanaOut$FDR.DE <- p.adjust(goanaOut$P.DE, method="BH")
topGOarray <- topGO(goanaOut, ontology="BP", number=50)
topGOarray
```
```{r}
goanaOut_BP <- goanaOut[goanaOut$Ont == "BP",]
print(paste("Amount of significant GO Biological Process terms:", as.character(sum(goanaOut_BP$FDR.DE < 0.05))))
```
```{r}
topGOarray[order(topGOarray$N), ]
```
```{r}
write.table(LIMMAout_RMA, sep= "\t", file="C:/Users/Luca Visser/Documents/1st master/AHTA/Practica/Project/analysis/Comparison/limmaExprsArray_results.txt")
array_GSA_res <- topGO(goanaOut, ontology="BP", number=100)
write.table(array_GSA_res, sep= "\t", file="C:/Users/Luca Visser/Documents/1st master/AHTA/Practica/Project/analysis/Comparison/array_GSA_results.txt")
```

```{r include = FALSE}
# Generate pngs
png("mdsplot.png")
plotMDS(data.norm, top = 500, col=as.double(annot.filt$disease))
dev.off()

png("normboxplot.png")
boxplot(d2)
dev.off()

png("maplot.png")
limma::plotMA(fit2, status = p, hl.cex = 0.1)
abline(h=c(1,-1), col=c("blue", "blue"), lty=c(2,2))
dev.off()

png("volcanoplot.png")
volcanoplot(fit2)
dev.off()

# Generate tables
library(xtable)
topHits <- LIMMAout_RMA[1:10,]
GeneSymbols <- mapIds(org.Hs.eg.db, gsub('.*\\.', '', rownames(topHits)), 'SYMBOL', 'ENSEMBL')
rownames(topHits) <- GeneSymbols
xtable(topHits, type = "latex", file = "limmat10.tex", digits = c(2, 2, 2, 2, -2, -2, 2))
```





