---
title: 'Group 1: RNA-seq (coding) analysis AHTA'
author: 
- Luca Visser
date: 2022
output:
  pdf_document:
    dev: png
    toc: true
    latex_engine: xelatex
  html_document:
    df_print: paged
    toc: true
header-includes: \setlength{\emergencystretch}{3em}
---

```{r global_options, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(fig.align = 'center')
knitr::opts_chunk$set(out.width = '60%')
knitr::opts_chunk$set(fig.width=12, fig.height=8)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

```{r}
suppressPackageStartupMessages({
  library(ggplot2)
  library(htmltools)
  library(biomaRt)
  library(tximport)
  library(dplyr)
  library(GEOquery)
  library(annotation)
  library(edgeR)
  library(limma)
  library(SummarizedExperiment)
})
```

# General info

Platform type: Illumina NextSeq 500 

The GEO dataset with accession ID GSE159699 (expression profiling by high-throughput sequencing sequencing of human postmortal hippocampal brain tissues) will be analysed in a case versus control framework for differential gene expression. The sequencing was done with single ends.

#Pipeline

## Import gene counts and metadata

```{r}
id <- "GSE159699"

gse <- getGEO(id)
gse <- gse[[1]]

annot <- pData(gse)
```

```{r eval = FALSE}
unlink(id, recursive = T)
filePath <- rownames(getGEOSuppFiles(id))

for (file in list.files(id, full.names = T)){
  gunzip(file)
}
```

```{r}
counts <- read.delim(list.files(id, full.names = T))

library('org.Hs.eg.db')
ENSIDs <- mapIds(org.Hs.eg.db, counts$refGene, 'ENSEMBL', 'SYMBOL')
counts <- counts[!(duplicated(ENSIDs) | is.na(ENSIDs)),]
rownames(counts) <- ENSIDs[!(duplicated(ENSIDs) | is.na(ENSIDs))]
counts <- dplyr::select(counts, -refGene)
```

```{r}
annot$disease <- "CT"
annot$disease[grep("AD", annot$title)] <- "AD"
annot$disease <- as.factor(annot$disease)

print(annot)
```

## Differential expression analysis

```{r}
dge <- DGEList(counts)
colnames(dge) <- substring(gsub('\\.', '-', colnames(dge)), 2)
annot <- annot[order(annot$disease), ]

design <- model.matrix(~disease, annot)
keep <- filterByExpr(dge, design)
table(keep)
dge <- dge[keep, ,keep.lib.sizes=FALSE]

dge <- calcNormFactors(dge)
```

```{r}
plotMDS(dge, top = 500, col=as.double(annot$disease))
```
```{r}
dge <- estimateDisp(dge, design, robust=TRUE)
plotBCV(dge)
```
```{r}
fit = glmQLFit(dge,design)
LRT <- glmQLFTest(fit)
topTags(LRT)
```
```{r}
summary(dt <- decideTestsDGE(LRT))
```
```{r}
plotSmear(LRT,de.tags=rownames(dge)[as.logical(dt)])
abline(h=c(1,-1), col=c("blue", "blue"), lty=c(2,2))
```
```{r}
hist(LRT$table$PValue, xlab="P-value", main="Histogram of P-values")
```
```{r}
FDR <- p.adjust(LRT$table$PValue, "fdr")
hist(FDR)
```
```{r}
LRT$table$FDR <- FDR
sign_genes <- LRT$table[FDR<0.05,]
sign_genes <- sign_genes[order(sign_genes$FDR),]

topHits <- LRT$table[order(LRT$table$FDR, decreasing=F),]
GeneSymbols <- mapIds(org.Hs.eg.db, rownames(topHits), 'SYMBOL', 'ENSEMBL')

topHits <- topHits[1:10,c("logFC", "FDR")]
topHits
```
## Gene set analysis

```{r}
EntrezIDs <- mapIds(org.Hs.eg.db, rownames(sign_genes), 'ENTREZID', 'ENSEMBL')

#subset for non duplicated and mapped genes
sign_genes_entrez <- sign_genes[!(duplicated(EntrezIDs) | is.na(EntrezIDs)),]
#make rownames the Entrez gene ID
rownames(sign_genes_entrez) <- EntrezIDs[!(duplicated(EntrezIDs) | is.na(EntrezIDs))]
```
```{r}
library(limma)
goanaOut <- goana(de=rownames(sign_genes_entrez), species="Hs", trend=T)
goanaOut <- goanaOut[order(goanaOut$P.DE, decreasing=FALSE),]
goanaOut$FDR.DE <- p.adjust(goanaOut$P.DE, method="BH")
topGORNA <- topGO(goanaOut, ontology="BP", number=50)
topGORNA
```
```{r}
goanaOut_BP <- goanaOut[goanaOut$Ont == "BP",]
print(paste("Amount of significant GO Biological Process terms:",
as.character(sum(goanaOut_BP$FDR.DE < 0.05))))
```

```{r}
topGORNA[order(topGORNA$N),]
```
```{r}
edgeR_res <- LRT$table[order(LRT$table$FDR, decreasing=F),]
write.table(edgeR_res, sep= "\t", file="C:/Users/Luca Visser/Documents/1st master/AHTA/Practica/Project/analysis/Comparison/RNAseq_results.txt")
RNAseq_GSA_res <- topGO(goanaOut, ontology="BP", number=100)
write.table(RNAseq_GSA_res, sep= "\t", file="C:/Users/Luca Visser/Documents/1st master/AHTA/Practica/Project/analysis/Comparison/RNAseq_GSA_results.txt")
```

```{r include = FALSE}
# Generate pngs
png("mdsplot.png")
plotMDS(dge, top = 500, col=as.double(annot$disease))
abline(h=c(1,-1), col=c("blue", "blue"), lty=c(2,2))
dev.off()

png("bcvplot.png")
plotBCV(dge)
dev.off()

png("smearplot.png")
plotSmear(LRT,de.tags=rownames(dge)[as.logical(dt)])
dev.off()

# Generate tables
library(xtable)
topHits <- LRT$table[order(LRT$table$FDR, decreasing=F),]
topHits <- topHits[1:10,]
GeneSymbols <- mapIds(org.Hs.eg.db, rownames(topHits), 'SYMBOL', 'ENSEMBL')
rownames(topHits) <- GeneSymbols
xtable(as.data.frame(topHits), type = "latex", file = "limmat10.tex", digits = c(2, 2, 2, 2, -2, -2))
```

















